db/import_pindex.go

```diff
@@ -0,0 +1,80 @@
+package db
+
+import (
+	"expvar"
+	"os"
+
+	"github.com/couchbase/cbgt"
+	"github.com/couchbase/sync_gateway/base"
+	"github.com/pkg/errors"
+)
+
+// init registers the PIndex type definition.  This is invoked by cbgt when a Pindex (collection of
+// vbuckets) is assigned to this node.
+func (il *importListener) RegisterImportPindexImpl() {
+
+	base.Infof(base.KeyDCP, "Registering PindexImplType for %s", base.CBGTIndexTypeSyncGatewayImport)
+	cbgt.RegisterPIndexImplType(base.CBGTIndexTypeSyncGatewayImport,
+		&cbgt.PIndexImplType{
+			New:       il.NewImportPIndexImpl,
+			Open:      il.OpenImportPIndexImpl,
+			OpenUsing: il.OpenImportPIndexImplUsing,
+			Description: "general/syncGateway-import " +
+				" - import processing for shared bucket access",
+		})
+}
+
+// NewImportPIndexImpl is called when the node is first added to the cbgt cfg.  On a node restart,
+// OpenImportPindexImpl is called, and indexParams aren't included.
+func (il *importListener) NewImportPIndexImpl(indexType, indexParams, path string, restart func()) (cbgt.PIndexImpl, cbgt.Dest, error) {
+
+	// TODO: Would really rather not require any file persistence here
+	// https://issues.couchbase.com/browse/MB-36085
+	base.Infof(base.KeyDCP, "NewImportPindexImpl - indexType %s, path %s", indexType, path)
+
+	// Create the pindex-specific path
+	err := os.MkdirAll(path, 0700)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	importDest, err := il.NewImportDest()
+	if err != nil {
+		base.Errorf(base.KeyDCP, "Error creating NewImportDest during NewImportPIndexImpl: %v", err)
```

Comment:
Advisory:

These errors won't be seen on the console if the user doesn't have the DCP log key enabled.



I've generally gone with `KeyAll` for warnings and errors - although maybe we could think about ignoring log key filtering for warn level and up in the future.

```suggestion
		base.Errorf(base.KeyAll, "Error creating NewImportDest during NewImportPIndexImpl: %v", err)
```

Reply:
Makes sense, although tends to make me think we should just be removing the key parameter from Errorf and Warnf if this is the desired usage.

Reply:
That's a fair point. I'm on board with that instead of the slightly magic ignoring log key behaivour I suggested above. I'll file an issue.